# Segmenting and Clustering Neighbourhoods in Toronto
## Written by MHK

Welcome to the assignmnet. The objective is to fetch content pertaining to neighbourhoods in Toronto from Wikipedia in an order to cluster them into similar clusters.

The first step in doing so is to import the necessary libraries.
1. Numpy and Pandas for standard data manipulation
2. Json to handle JSON files
3. Wikipedia library to scrape tables from wikipedia
4. Matplotlib as the choice of visualization library
5. Scikit Learn as the library for the clustering algorithms
6. Requests library to handle url based requests
7. Geocoder for converting addresses to latitude and longitude values
8. Deepcopy for copying objects
9. Folium for drawing maps

NOTE: Some of the libraries might need to be installed depending on where you are running the code from.

### 1. IMPORTING THE NECESSARY LIBRARIES

In [37]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge wikipedia --yes #install wikipedia library
import wikipedia as wp

#!conda install -c conda-forge geocoder --yes #install geocoder
import geocoder

#deepcopy tool to avoid trickle effects
import copy

#folium for maping the neighbourhoods #!conda install -c conda-forge folium=0.5.0 --yes
import folium

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/DSX-Python35

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    altair-2.2.2               |           py35_1         462 KB  conda-forge
    branca-0.3.1               |             py_0          25 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         560 KB

The following NEW packages will be INSTALLED:

    altair:  2.2.2-py35_1 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forge


vincent-0.4.4        | 28 KB     | #############################

### 2. Scrape the Wikipedia Table

In [2]:
#Get the html source
#syntax for fetching:-> html = wp.page("wikipedia table name").html().encode("UTF-8")
html = wp.page("List of postal codes of Canada: M").html().encode("UTF-8")

#read it into a pandas df
df = pd.read_html(html, header = 0)[0]
print (df.columns) #confirm the correct labels
df.head(10) #initial df for comparing later steps

Index(['Postcode', 'Borough', 'Neighbourhood'], dtype='object')


Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M5A  Downtown Toronto       Regent Park
6      M6A        North York  Lawrence Heights
7      M6A        North York    Lawrence Manor
8      M7A      Queen's Park      Not assigned
9      M8A      Not assigned      Not assigned

### 3. Clean up the Table

In [3]:
#select only those rows where Borough isnot 'Not Assigned'
df = df[df['Borough']!="Not assigned"]

#if Neighbourhood is 'Not Assigned', copy Borough name over as the Neighbourhood name 
#assumes that any row with both neighbourhood and borough 'Not Assigned' has already been dropped
for i, neigh in enumerate(df['Neighbourhood']):
    if neigh == "Not assigned":
        df.iloc[i,2] = df.iloc[i,1]

#reset indices 
df.reset_index(drop=True, inplace=True)
df.head(10)

Postcode           Borough     Neighbourhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights
5      M6A        North York    Lawrence Manor
6      M7A      Queen's Park      Queen's Park
7      M9A         Etobicoke  Islington Avenue
8      M1B       Scarborough             Rouge
9      M1B       Scarborough           Malvern

In [4]:
#create a new copy of df 
df1 = copy.deepcopy(df)
df1.head()
#Add a spacer (,_) in front of the Neighbourhood name
df1['Neighbourhood'] = df['Neighbourhood']+', '
#print(df1.head(20))
#Groupby Postcode and Borough so that any Neighourhoods sharing Postcode and concatenated. The comma will serve as a separator
df1 = df1.groupby(['Postcode','Borough']).sum()
#Remove the last two characters which is the spacer we filled
df1['Neighbourhood'] = df1['Neighbourhood'].astype(str).str[:-2]
#reset indices
df1.reset_index(inplace=True)
df1.head()

Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

In [5]:
print(df1.shape) #print the df shape
df1.head()

(103, 3)


Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

In [6]:
df['Postcode'].nunique() #Number of unique postcodes - must match df.shape[0]

103

### 4. Get the postal codes from the CSV file shared in the notes

NOTE: Unfortunately, the geocoder just didn't work

In [7]:
#download the csv file with postcodes
!wget -q -O 'postal_codes.csv' http://cocl.us/Geospatial_data
print('Data downloaded!')

Data downloaded!


In [8]:
#copy into a dataframe
post_codes = pd.read_csv('postal_codes.csv')

In [29]:
result = pd.concat([df1, post_codes], axis=1)
result.head()

Postcode      Borough                           Neighbourhood Postal Code  \
0      M1B  Scarborough                          Rouge, Malvern         M1B   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union         M1C   
2      M1E  Scarborough       Guildwood, Morningside, West Hill         M1E   
3      M1G  Scarborough                                  Woburn         M1G   
4      M1H  Scarborough                               Cedarbrae         M1H   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

In [30]:
#Ensure that the Postcode of df file match the Postal Code of post_codes file
check = result["Postcode"] == result["Postal Code"]
check.value_counts()

True    103
dtype: int64

In [31]:
#drop the Postal Code column
result.drop(["Postal Code"],axis=1, inplace=True)
result.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

### 5. Let's visualize the neighbourhoods on a MAP

In [34]:
#Get the average lat and long values for the map center later
latitude = result['Latitude'].mean()
longitude = result['Longitude'].mean()
print(latitude, longitude)

43.70460773398059 -79.39715291165048


In [75]:
# create map of ONTARIO using latitude and longitude values
map_ONTARIO = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(result['Latitude'], result['Longitude'], result['Borough'], result['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_ONTARIO)  
    
map_ONTARIO

### 6. Get Neighbourhood info from Foursquare

In [39]:
#prepare credentials for Foursquare API
CLIENT_ID = 'KBHGEAWTOWHNWHTNESPPMLXCHANDDQQ0CCQXI1HLFAM2KDSZ' # your Foursquare ID
CLIENT_SECRET = 'G33KZ24X4TRWSYTHN1E1MZUGMMPHXJTLNYE3YYHOO2KKVSF4' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: KBHGEAWTOWHNWHTNESPPMLXCHANDDQQ0CCQXI1HLFAM2KDSZ
CLIENT_SECRET:G33KZ24X4TRWSYTHN1E1MZUGMMPHXJTLNYE3YYHOO2KKVSF4


In [50]:
result.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

#### Foursquare Time
Now let us play around with the Foursquare resource. We will test out with one neighbourhood to ensure the code is working.

Of course, objective is to get info of the nearby venues of those neighbourhoods.

In [56]:
#Test with one neighbourhood first
neighbourhood_latitude = result.loc[5, 'Latitude'] # neighborhood latitude value
neighbourhood_longitude = result.loc[5, 'Longitude'] # neighborhood longitude value

neighbourhood_name = result.loc[5, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Scarborough Village are 43.7447342, -79.23947609999999.


In [73]:
#preparing URL for Foursquare query
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, VERSION, neighbourhood_latitude, neighbourhood_longitude, radius, LIMIT)

url

'https://api.foursquare.com/v2/venues/explore?&client_id=KBHGEAWTOWHNWHTNESPPMLXCHANDDQQ0CCQXI1HLFAM2KDSZ&client_secret=G33KZ24X4TRWSYTHN1E1MZUGMMPHXJTLNYE3YYHOO2KKVSF4&v=20180605&ll=43.7447342,-79.23947609999999&radius=500&limit=100'

In [58]:
#Let's get the results from FOURSQUARE
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5cec32ad6a60714939e75c52'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-5150a8dae4b045dfb6581c85-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/playground_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d1e7941735',
         'name': 'Playground',
         'pluralName': 'Playgrounds',
         'primary': True,
         'shortName': 'Playground'}],
       'id': '5150a8dae4b045dfb6581c85',
       'location': {'cc': 'CA',
        'country': 'Canada',
        'distance': 41,
        'formattedAddress': ['Canada'],
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.74508851212816,
          'lng': -79.239335687338}],
        'lat': 43.74508851212816,
        'lng': -79.239335687338},
      

In [59]:
# function that extracts the category of the venue (copied from the Foursquare Lab - this will be used in the batch processor!)
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

#### Let's clean up the json file to narrow down the venues which is the only thing we are interested in

In [61]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0                     McCowan Park         Playground  43.745089 -79.239336
1  True Discount Convenience Store  Convenience Store  43.743854 -79.244650

### OBSERVATION
Seems like not a very populated neighbourhood. Interestingly, the FOURSQUARE query also returned a warning saying there isn't much happening around the neighbourhood

#### NEXT - we rope in the same info and fetch the categories of venues for all the neighbourhood using this batch processor!

In [62]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### CALL IT!

In [65]:
ONTARIO_venues = getNearbyVenues(names=result['Neighbourhood'],
                                   latitudes=result['Latitude'],
                                   longitudes=result['Longitude']
                                  )


Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

In [68]:
print(ONTARIO_venues.shape)
ONTARIO_venues.head()

(2236, 7)


Neighbourhood  Neighbourhood Latitude  \
0                          Rouge, Malvern               43.806686   
1  Highland Creek, Rouge Hill, Port Union               43.784535   
2       Guildwood, Morningside, West Hill               43.763573   
3       Guildwood, Morningside, West Hill               43.763573   
4       Guildwood, Morningside, West Hill               43.763573   

   Neighbourhood Longitude                            Venue  Venue Latitude  \
0               -79.194353                          Wendy's       43.807448   
1               -79.160497            Royal Canadian Legion       43.782533   
2               -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
3               -79.188711                G & G Electronics       43.765309   
4               -79.188711                       Marina Spa       43.766000   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.189914           Pizza Place  
3       -79.191537     Electronics Store  
4       -79.191000                   Spa

In [74]:
# Let's have a look at the number of venues around each neighbourhood
ONTARIO_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide, King, Richmond                                               100   
Agincourt                                                                5   
Agincourt North, L'Amoreaux East, Milliken, Ste...                       2   
Albion Gardens, Beaumond Heights, Humbergate, J...                      11   
Alderwood, Long Branch                                                  10   
Bathurst Manor, Downsview North, Wilson Heights                         18   
Bayview Village                                                          4   
Bedford Park, Lawrence Manor East                                       24   
Berczy Park                                                             56   
Birch Cliff, Cliffside West                                              4   
Bloordale Gardens, Eringate, Markland Wood, Old...                       7   
Brockton, Exhibition Place, Parkdale Village                            19   
Business Reply Mail Processing Centre 969 Eastern                       16   
CFB Toronto, Downsview East                                              2   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             40   
Caledonia-Fairbanks                                                      5   
Canada Post Gateway Processing Centre                                   11   
Cedarbrae                                                                8   
Central Bay Street                                                      85   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                16   
Church and Wellesley                                                    87   
Clairlea, Golden Mile, Oakridge                                         10   
Clarks Corners, Sullivan, Tam O'Shanter                                 11   
Cliffcrest, Cliffside, Scarborough Village West                          2   
Cloverdale, Islington, Martin Grove, Princess G...                       1   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              36   
Davisville North                                                         7   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                      15   
Del Ray, Keelesdale, Mount Dennis, Silverthorn                           4   
Design Exchange, Toronto Dominion Centre                               100   
Don Mills North                                                          5   
Dorset Park, Scarborough Town Centre, Wexford H...                       6   
Dovercourt Village, Dufferin                                            20   
Downsview Central                                                        4   
Downsview Northwest                                                      4   
Downsview West                                                           4   
Downsview, North Park, Upwood Park                                       4   
East Birchmount Park, Ionview, Kennedy Park                              4   
East Toronto                                                             5   
Emery, Humberlea                                                         2   
Fairview, Henry Farm, Oriole                                            65   
First Canadian Place, Underground city                                 100   
Flemingdon Park, Don Mills South                                        22   
Forest Hill North, Forest Hill West                                      5   
Glencairn                                                                5   
Guildwood, Morningside, West Hill                                        9   
Harbord, University of Toronto                                          36

In [71]:
print('There are {} uniques categories.'.format(len(ONTARIO_venues['Venue Category'].unique())))

There are 276 uniques categories.


## OBSERVATIONS
1. A lot of neighbourhoods do not have many venues in their vicinity. They seem to be far from the city. Might be useful to remove them as they might not be good places to live anyway (from this assignment's perspective). Would be intersting if the clustering algorithm would maybe group them together? Let's see.
2. 276 unique categories, although I'm sure there are a lot of redundant categories e.g. cafe and coffee shop. Might be useful to remove those redundancies (for future reference).


### 7. Analyze Each Neighborhood

In [77]:
# one hot encoding
ON_onehot = pd.get_dummies(ONTARIO_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
ON_onehot['Neighbourhood'] = ONTARIO_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [ON_onehot.columns[-1]] + list(ON_onehot.columns[:-1])
ON_onehot = ON_onehot[fixed_columns]

print(ON_onehot.shape) #should have 1 column for every unique category
ON_onehot.head()

(2236, 277)


Neighbourhood  Accessories Store  Adult Boutique  \
0                          Rouge, Malvern                  0               0   
1  Highland Creek, Rouge Hill, Port Union                  0               0   
2       Guildwood, Morningside, West Hill                  0               0   
3       Guildwood, Morningside, West Hill                  0               0   
4       Guildwood, Morningside, West Hill                  0               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    1               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Burger 

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [80]:
ON_grouped = ON_onehot.groupby('Neighbourhood').mean().reset_index()

print(ON_grouped.shape) #size[1] should match
ON_grouped

(99, 277)


Neighbourhood  Accessories Store  \
0                            Adelaide, King, Richmond           0.000000   
1                                           Agincourt           0.000000   
2   Agincourt North, L'Amoreaux East, Milliken, St...           0.000000   
3   Albion Gardens, Beaumond Heights, Humbergate, ...           0.000000   
4                              Alderwood, Long Branch           0.000000   
5     Bathurst Manor, Downsview North, Wilson Heights           0.000000   
6                                     Bayview Village           0.000000   
7                   Bedford Park, Lawrence Manor East           0.000000   
8                                         Berczy Park           0.000000   
9                         Birch Cliff, Cliffside West           0.000000   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...           0.000000   
11       Brockton, Exhibition Place, Parkdale Village           0.000000   
12  Business Reply Mail Processing Centre 969 Eastern           0.000000   
13                        CFB Toronto, Downsview East           0.000000   
14  CN Tower, Bathurst Quay, Island airport, Harbo...           0.000000   
15                        Cabbagetown, St. James Town           0.000000   
16                                Caledonia-Fairbanks           0.000000   
17              Canada Post Gateway Processing Centre           0.000000   
18                                          Cedarbrae           0.000000   
19                                 Central Bay Street           0.000000   
20          Chinatown, Grange Park, Kensington Market           0.000000   
21                                           Christie           0.000000   
22                               Church and Wellesley           0.000000   
23                    Clairlea, Golden Mile, Oakridge           0.000000   
24            Clarks Corners, Sullivan, Tam O'Shanter           0.000000   
25    Cliffcrest, Cliffside, Scarborough Village West           0.000000   
26  Cloverdale, Islington, Martin Grove, Princess ...           0.000000   
27                     Commerce Court, Victoria Hotel           0.000000   
28                                         Davisville           0.000000   
29                                   Davisville North           0.000000   
30  Deer Park, Forest Hill SE, Rathnelly, South Hi...           0.000000   
31     Del Ray, Keelesdale, Mount Dennis, Silverthorn           0.000000   
32           Design Exchange, Toronto Dominion Centre           0.000000   
33                                    Don Mills North           0.000000   
34  Dorset Park, Scarborough Town Centre, Wexford ...           0.000000   
35                       Dovercourt Village, Dufferin           0.000000   
36                                  Downsview Central           0.000000   
37                                Downsview Northwest           0.000000   
38                                     Downsview West           0.000000   
39                 Downsview, North Park, Upwood Park           0.000000   
40        East Birchmount Park, Ionview, Kennedy Park           0.000000   
41                                       East Toronto           0.000000   
42                                   Emery, Humberlea           0.000000   
43                       Fairview, Henry Farm, Oriole           0.015385   
44             First Canadian Place, Underground city           0.000000   
45                   Flemingdon Park, Don Mills South           0.000000   
46                Forest Hill North, Forest Hill West           0.000000   
47                                          Glencairn           0.000000   
48                  Guildwood, Morningside, West Hill           0.000000   
49                     Harbord, University of Toronto           0.000000   
50  Harbourfront East, Toronto Islands, Union Station           0.000000   
51                          Harbourfront, Regent Park           0.000000   
52          

#### Let's print each neighborhood along with the top 5 most common venues

In [83]:
num_top_venues = 5

for hood in ON_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = ON_grouped[ON_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2      Thai Restaurant  0.04
3                  Bar  0.04
4  American Restaurant  0.04


----Agincourt----
            venue  freq
0    Skating Rink   0.2
1  Breakfast Spot   0.2
2      Print Shop   0.2
3          Lounge   0.2
4  Sandwich Place   0.2


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                venue  freq
0          Playground   0.5
1                Park   0.5
2  Miscellaneous Shop   0.0
3       Movie Theater   0.0
4               Motel   0.0


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                  venue  freq
0         Grocery Store  0.18
1              Pharmacy  0.09
2           Pizza Place  0.09
3  Fast Food Restaurant  0.09
4            Beer Store  0.09


----Alderwood, Long Branch----
                venue  freq
0         Pizza Place   0.2
1    

#### Let's put that into a *pandas* dataframe
First, let's write a function to sort the venues in descending order.

In [84]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [179]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = ON_grouped['Neighbourhood']

for ind in np.arange(ON_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(ON_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt            Print Shop   
2  Agincourt North, L'Amoreaux East, Milliken, St...            Playground   
3  Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse       Thai Restaurant   
1                Lounge          Skating Rink        Breakfast Spot   
2                  Park           Yoga Studio   Dumpling Restaurant   
3              Pharmacy   Fried Chicken Joint           Pizza Place   
4              Pharmacy                   Gym           Coffee Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar   American Restaurant        Cosmetics Shop   
1        Sandwich Place           Yoga Studio        Discount Store   
2                 Diner        Discount Store               Dog Run   
3           Coffee Shop        Discount Store        Sandwich Place   
4          Skating Rink    Athletics & Sports                  Pool   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Bakery                 Hotel           Burger Joint  
1               Dog Run      Doner Restaurant             Donut Shop  
2      Doner Restaurant            Donut Shop              Drugstore  
3            Beer Store   Japanese Restaurant   Fast Food Restaurant  
4        Sandwich Place                   Pub         Discount Store

## 8. Cluster Neighborhoods

In [180]:
ON_grouped.head()

Neighbourhood  Accessories Store  \
0                           Adelaide, King, Richmond                0.0   
1                                          Agincourt                0.0   
2  Agincourt North, L'Amoreaux East, Milliken, St...                0.0   
3  Albion Gardens, Beaumond Heights, Humbergate, ...                0.0   
4                             Alderwood, Long Branch                0.0   

   Adult Boutique  Afghan Restaurant  Airport  Airport Food Court  \
0             0.0                0.0      0.0                 0.0   
1             0.0                0.0      0.0                 0.0   
2             0.0                0.0      0.0                 0.0   
3             0.0                0.0      0.0                 0.0   
4             0.0                0.0      0.0                 0.0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0           0.0             0.0              0.0               0.0   
1           0.0             0.0              0.0               0.0   
2           0.0             0.0              0.0               0.0   
3           0.0             0.0              0.0               0.0   
4           0.0             0.0              0.0               0.0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                 0.04           0.0       0.0         0.01        0.01   
1                 0.00           0.0       0.0         0.00        0.00   
2                 0.00           0.0       0.0         0.00        0.00   
3                 0.00           0.0       0.0         0.00        0.00   
4                 0.00           0.0       0.0         0.00        0.00   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                  0.0              0.02                 0.0          0.0   
1                  0.0              0.00                 0.0          0.0   
2                  0.0              0.00                 0.0          0.0   
3                  0.0              0.00                 0.0          0.0   
4                  0.0              0.00                 0.1          0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank   Bar  \
0            0.0        0.0         0.0         0.0    0.03   0.0  0.04   
1            0.0        0.0         0.0         0.0    0.00   0.0  0.00   
2            0.0        0.0         0.0         0.0    0.00   0.0  0.00   
3            0.0        0.0         0.0         0.0    0.00   0.0  0.00   
4            0.0        0.0         0.0         0.0    0.00   0.0  0.00   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0             0.0               0.0               0.0                 0.0   
1             0.0               0.0               0.0                 0.0   
2             0.0               0.0               0.0                 0.0   
3             0.0               0.0               0.0                 0.0   
4             0.0               0.0               0.0                 0.0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0    0.0              0.0       0.0    0.000000                 0.0   
1    0.0              0.0       0.0    0.000000                 0.0   
2    0.0              0.0       0.0    0.000000                 0.0   
3    0.0              0.0       0.0    0.090909                 0.0   
4    0.0              0.0       0.0    0.000000                 0.0   

   Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0        0.0     0.0            0.0       0.01       0.0   
1        0.0     0.0            0.0       0.00       0.0   
2        0.0     0.0            0.0       0.00       0.0   
3        0.0     0.0            0.0       0.00       0.0   
4        0.0     0.0            0.0       0.00       0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                  0.01            0.02      0.0          0.0   
1     

In [189]:

# set number of clusters
kclusters = 3

ON_grouped_clustering = ON_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(ON_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

(99, 16)


array([0, 0, 2, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [182]:
result.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [183]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

ON_merged = result

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
ON_merged = ON_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

ON_merged.head() # check the last columns!

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             0.0  Fast Food Restaurant   Dumpling Restaurant   
1 -79.160497             0.0                   Bar           Yoga Studio   
2 -79.188711             0.0     Electronics Store           Pizza Place   
3 -79.216917             0.0           Coffee Shop     Indian Restaurant   
4 -79.239476             0.0                Lounge      Hakka Restaurant   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                        Diner        Discount Store               Dog Run   
1  Eastern European Restaurant               Dog Run      Doner Restaurant   
2                          Spa   Rental Car Location        Breakfast Spot   
3            Korean Restaurant           Yoga Studio   Dumpling Restaurant   
4                       Bakery    Athletics & Sports  Caribbean Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Doner Restaurant            Donut Shop             Drugstore   
1            Donut Shop             Drugstore   Dumpling Restaurant   
2        Medical Center          Intersection    Mexican Restaurant   
3        Discount Store               Dog Run      Doner Restaurant   
4       Thai Restaurant                  Bank   Fried Chicken Joint   

         9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant                  Field  
1            Electronics Store                  Diner  
2                  Yoga Studio             Donut Shop  
3                   Donut Shop              Drugstore  
4                    Drugstore       Doner Restaurant

In [184]:
ON_merged['Cluster Labels'].value_counts()

0.0    84
1.0    12
2.0     3
Name: Cluster Labels, dtype: int64

NOTE: Seems like there are some NaN values which have crept in after the clustering.

In [185]:
ON_merged.shape

(103, 16)

In [186]:
#Dropping those rows which give NaN clusters
ON_merged = ON_merged.dropna()
ON_merged['Cluster Labels'] = ON_merged['Cluster Labels'].astype(int)
ON_merged.shape

(99, 16)

In [187]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(ON_merged['Latitude'], ON_merged['Longitude'], ON_merged['Neighbourhood'], ON_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters